server-common/src/main/java/org/apache/kafka/server/purgatory/DelayedOperation.java

```diff
@@ -51,32 +51,15 @@ public DelayedOperation(long delayMs) {
      *
      * 1. The operation has been verified to be completable inside tryComplete()
      * 2. The operation has expired and hence needs to be completed right now
-     *
-     * Return true iff the operation is completed by the caller: note that
-     * concurrent threads can try to complete the same operation, but only
-     * the first thread will succeed in completing the operation and return
-     * true, others will still return false.
+     * 
      */
-    public boolean forceComplete() {
+    public void forceComplete() {
         // Do not proceed if the operation is already completed.
-        if (completed) {
-            return false;
-        }
-        // Attain lock prior completing the request.
-        lock.lock();
-        try {
-            // Re-check, if the operation is already completed by some other thread.
-            if (!completed) {
-                completed = true;
-                // cancel the timeout timer
-                cancel();
-                onComplete();
-                return true;
-            } else {
-                return false;
-            }
-        } finally {
-            lock.unlock();
+        if (!completed) {
```

Comment:
Both callers already check completed. So, this seems redundant?

Reply:
In the `safeTryCompleteOrElse()` method, `tryComplete()` is called without first checking `isCompleted()`.
we could move the `isCompleted()` check into `safeTryCompleteOrElse()` and remove in `forceComplete()`.

Reply:
Removing the condition introduces another restriction: `forceComplete` must be called only once in `tryComplete`. That may be error-prone I think

Reply:
I agree, it seems to get more complicated after removing boolean and that's what I was trying to explain [here](https://github.com/apache/kafka/pull/19759#discussion_r2104086816).